LFS-444: New question type: chromosome

IntegralProgrammer opened the pull request:
- This PR creates a new question type `lfs:ChromosomeAnswer` which is displayed as a drop-down menu allowing for the selection of chromosomes `1..N` and optionally `X`, `Y`, `Z`, `W`, `MT`

- To test, build the branch **LFS-444_test** and create a new *Demographics* form. Options can be adjusted in the `SLING-INF/content/Demographics.json` file.

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Add lfs:ChromosomeAnswer and associated question properties to dataentry.cnd

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Add simplified implementation of ChromosomeQuestion.jsx which functions simply as text entry but saves as data type lfs:ChromosomeAnswer

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Modify MultipleChoice.jsx so that if the dataType is chromosome, a React <Select> menu will be created listing chromosomes 1-22

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Should display only 8 chromosomes, but displays 22 instead

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Move default of 22 chromosomes to an inline default instead of a prop default

This fixes the error of always 22 chromosomes being displayed regardless of the chromosomeNumber property in the JSON file

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Add support for X chromosome

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Add support for Y chromosome

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Add support for Z chromosome

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Add support for W chromosome

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Add support for MT (mitochondria)

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Refactor: pass in displayType from ChromosomeQuestion.jsx instead of deriving it from 'chromosome' data type in MultipleChoice.jsx

IntegralProgrammer committed:
LFS-444: New question type: chromosome
Move all logic specific to chromosomes from MultipleChoice.jsx into ChromosomeQuestion.jsx

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ChromosomeQuestion.jsx:
```diff
@@ -0,0 +1,124 @@
+      menuitems_list.push("MT");
+  } else {
+    chromosomeMT = false;
+  }
```
This looks like "boring" code, too much internal similarity in a too long portion of code. Should be refactored in a more generic way.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -32,8 +32,8 @@ const IS_DEFAULT_POS = 2;
 function MultipleChoice(props) {
-  let { classes, existingAnswer, ghostAnchor, input, textbox, onChange, additionalInputProps, muiInputProps, error, ...rest } = props;
-  let { maxAnswers, minAnswers } = {...props.questionDefinition, ...props};
+  let { classes, existingAnswer, ghostAnchor, input, textbox, menuitems, onChange, additionalInputProps, muiInputProps, error, ...rest } = props;
```
I think there's no need for `menuitems`, `defaults` does the same.

marta- commented on questionnaires/clinical-data/src/main/resources/SLING-INF/content/Demographics.json:
```diff
@@ -5,6 +5,13 @@
     "jcr:primaryType": "lfs:Question",
     "text": "Patient ID"
   },
+  "chromosome1": {
```
Our best practice for facilitating testing the PR by reviewers is to create another branch with the suffix `_test` (e.g `LFS-444_test`) containing all the changes from the PR branch (LFS-444) plus any modifications that make testing straightforward, such as new questions/questionnaires, but that should not end up in the codebase. This way the PR will be easy to merge without the need to clean up test code.

IntegralProgrammer replied:
I have reset the `Demographics.json` file in this `LFS-444` branch to the version in the `dev` branch and a new `LFS-444_test` branch holds a version of `Demographics.json` containing a chromosome question. :heavy_check_mark:

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/ChromosomeQuestion.jsx:
```diff
@@ -0,0 +1,124 @@
+function ChromosomeQuestion(props) {
+  let { ...rest } = props;
+  let { chromosomeNumber,  chromosomeX, chromosomeY, chromosomeZ, chromosomeW, chromosomeMT } = {...props.questionDefinition, ...props};
+  chromosomeNumber = chromosomeNumber || 22;
```
@OrdiNeu @sdumitriu do you know of any other more elegantoptions for defining defaults for a data type? Hardcoding them in the UI component seems like a suboptimal choice.

sdumitriu committed:
LFS-444: New question type: chromosome
More documentation

sdumitriu committed:
LFS-444: New question type: chromosome
There's no need for defining `rest` if it's the same as `props`

sdumitriu committed:
LFS-444: New question type: chromosome
Remove wrongly used props

sdumitriu committed:
LFS-444: New question type: chromosome
There's no need for a new `displayType` parameter, we already have `displayMode`

sdumitriu committed:
LFS-444: New question type: chromosome
Don't add a new `menuitems` prop, there's `defaults` for that already

sdumitriu committed:
LFS-444: New question type: chromosome
More consistent styling of the new input

sdumitriu committed:
LFS-444: New question type: chromosome
Fixed React warning about unkeyed list

sdumitriu committed:
LFS-444: New question type: chromosome
Tracking the selected option should be done using `setSelection`, not through the ghost input

sdumitriu committed:
LFS-444: New question type: chromosome
Made code more maintainable, avoiding a long list of if..else statements
